Feat/Scrum 138 아티스트 선택 화면 로직 구현

inwoodev opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경
- [기능 명세서](https://docs.google.com/spreadsheets/d/124j3y6raLcu2VXtSXAo17-TNqPJkCdMxp33xmS3ZHbA/edit?gid=1710828598#gid=1710828598) 에 따라
- 유저는 아티스트 이름을 통해 아티스트를 검색할 수 있어야 한다.
- 유저는 '나의 관심 아티스트'로 추가할 아티스트 버블을 선택할 수 있어야 한다.

## 목표
- 실시간 검색 기능 구현
- 아티스트 버블(cell) 노출을 무한 스크롤 방식으로 구현
- 나의 관심사 아티스트 선택 기능 구현
  - 최대 30개까지 선택가능하도록 제한

## 결과 (Optional)
- [x] 초기 아티스트 검색 기능 구현
  - 검색창이 비활성화 상태이며 아무런 검색어가 입력되지 않은 경우 초기 아티스트 데이터를 불러온다.
![Simulator Screen Recording - iPhone 15 Pro - 2024-07-21 at 21 40 55](https://github.com/user-attachments/assets/1c249626-9a43-4bac-8b5d-a5944b7f0fc6)

- [x] 실시간 검색 기능 구현
![Simulator Screen Recording - iPhone 15 Pro - 2024-07-21 at 21 38 37](https://github.com/user-attachments/assets/426f74e0-e6e6-4b62-9f51-cd51178d4fac)


- [x] 아티스트 목록을 무한 스크롤 방식으로 구현
![Simulator Screen Recording - iPhone 15 Pro - 2024-07-21 at 21 42 01](https://github.com/user-attachments/assets/5390a774-a19e-44ed-b50f-6f1a9f38b71f)

- [x] 나의 관심사 아티스트 선택 기능 구현
  - [x] 선택된 아티스트 정렬 기능 구현 
  - [x] 최대 30개까지 선택가능하도록 제한(초기 데이터가 별로 없는 관계로 예시에서는 5개로 제한 설정)
  - [x] 좋아하는 아티스트가 한 개라도 선택될 경우 완료 버튼 활성화

<p align="left">
  <img src="https://github.com/user-attachments/assets/8af0ba17-3e07-4b1b-bbea-0e9bf6d629b9" alt="Simulator Screen Recording - iPhone 15 Pro - 2024-07-21 at 21 45 28" width="300" />
  <img src="https://github.com/user-attachments/assets/c9f64b7a-033e-4254-a68f-0d62e1c79c5c" alt="Simulator Screen Recording - iPhone 15 Pro - 2024-07-21 at 21 47 16" width="300" />
</p>

inwoodev committed:
feat<Tuist>: DomainArtist 모듈 추가 및 Dependency 정의

inwoodev committed:
refactor: CombinCococa → Combine으로 네이밍 수정

inwoodev committed:
feat: Combine & Collection helper extension및 객체 추가

1. Collection Extension
- isNotEmpty추가

2. Extensions
- Publisher에 mapToResult 메서드 추가
- UIScrollView사용시 유저가 맨 아래까지 스크롤했을 때 이벤트를 방출하는 Publisher 추가
- UITextField에서 입력 및 입력 종료를 감지하여 이벤트를 방출하는 Publisher 추가

3. CombineKeyboard
- 화면 키보드 높이 변경을 감지하여 이벤트를 방출하는 객체 구현

inwoodev committed:
feat, refactor: 좋아하는 아티스트 화면 대응 FeelinAPI 수정 및 관련 Response 객체 구현

- FeelinAPI
  - getArtist, searchArtist, postFavoriteArtists api 추가

- Response
  - GetArtistResponse & EmptyResponse 추가

inwoodev committed:
feat: ArtistDomain 로직 구현

- ArtistAPIService
  - 아티스트 정보를 조회하고 검색하며, 좋아하는 아티스트를 등록하는 API 서비스

- ArtistPaginationService
  - 아티스트 목록 페이지네이션 관리를 위한 서비스. 검색어, 현재 페이지, 다음 페이지 존재 여부, 로딩 상태를 관리하며, 스레드 안전성을 보장하기 위해 NSLock을 사용합니다.

- GetArtistsUseCase
  - 아티스트 목록을 페이지 단위로 가져오고 페이지네이션 상태를 관리하는 유스케이스. API 요청, 로딩 상태 및 페이지 정보를 처리하여 아티스트 데이터를 반환합니다

- PostFavoriteArtistsUseCase
  - 아티스트를 즐겨찾기 목록에 추가하는 유스케이스. 주어진 아티스트 ID 목록을 서버에 전송하여 즐겨찾기 상태를 업데이트합니다.

- SearchArtistsUseCase
  - 키워드로 아티스트를 검색하고 페이지네이션 상태를 관리하는 유스케이스. 검색어에 따라 아티스트 목록을 API로 요청하고, 로딩 상태 및 페이지 정보를 업데이트합니다.

inwoodev committed:
feat, refactor: 좋아하는 아티스트 화면 로직 구현 및 관련 객체 수정

- 초기 아티스트 목록을 가져오는 기능 추가
  - TextField가 activate되지 않았으며 검색입력값이 없을 경우 초기 아티스트 목록을 가져오는 로직 구현
 - 실시간 아티스트 검색 기능 추가
  - SearchTextField에 값이 추가될 경우 1초의 debounce 이후 검색을 수행하는 로직 구현

- 좋아하는 아티스트 선택 기능 추가
  - 좋아하는 아티스트를 선택할 경우 리스트의 맨 앞으로 정렬되는 로직 구현(그 외 아티스트들은 id값으로 정렬(오름차순 정렬)
  - 30개이상 선택될 경우 에러 메세지를 띄우는 로직 추가

- 좋아하는 아티스트 선택 완료 버튼 탭할 때 서버로 해당 아티스트목록의 id를 전달하는 로직 구현

그외:
- Kingfisher를 활용하여 FeelinArtistCell에 이미지를 불러오도록 로직 수정
- 키보드가 화면에 떠있을 때/없을 때에 따라 collectionView와 section간에 inset을 주는 기능 추가
- 아티스트 셀의 이름이 짤리는 문제 해결
- CoreExport에 누락된 import 추가(CoreLocalStorage)

inwoodev committed:
refactor: DIContainer에 ArtistAPIService & ArtistPaginationService 등록

inwoodev committed:
refactor: 좋아하는 아티스트 화면 동작테스트를 위해 Example앱에 의존성 주입

inwoodev commented on Projects/Shared/Util/Sources/CombineCocoa/CombineKeyboard.swift:
```diff
@@ -0,0 +1,27 @@
+import Combine
+import UIKit
+
+public struct CombineKeyboard {
+    static public var keyboardHeightPublisher: AnyPublisher<CGFloat, Never> {
+        let willShow = NotificationCenter.default.publisher(for: UIApplication.keyboardWillShowNotification)
+            .map({ notification in
+                let keyboardFrame = notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect
+                
+                return keyboardFrame?.height ?? 0
+            })
+
+        let willHide = NotificationCenter.default.publisher(for: UIApplication.keyboardWillHideNotification)
+            .map { _ in CGFloat(0) }
+
+        
+        return Publishers.MergeMany(willShow, willHide)
+            .eraseToAnyPublisher()
+    }
+}
```
NotificationCenter 또는 Publisher의 extension으로 구현할까 하다 이렇게 CombineKeyboard로 별도로 구현해 놓는게 더 가독성 측면에서 나을 듯 하여 별도의 객체로 만들어두었습니다. @derrickkim0109  어떻게 생각하시나요~?

derrickkim0109 replied:
오 너무 좋습니다~!

inwoodev commented on Projects/Shared/Util/Sources/Extensions/Combine/UITextField+Combine.swift:
CombineCocoa에서 거의 복붙한 ControlProperty를 활용하여 textPublisher및 editEndPublisher를 구현하였습니다. CombineCocoa에서 구현한 ControlProperty는 확인하셨다시피 events들을 선택할 때 폭 넓고 필요에 따라 조절할 수 있다는 장점이 있기에 가져와 봤습니다.

예를 들어 return버튼을 탭하는 것 뿐 아니라 closebutton에도 대응하기 위해서 `[.editingDidEnd, .editingDidEndOnExit]` 요 값들만 넣어주면 되기에 사용하기 매우 편리합니다~

inwoodev commented on Projects/Domain/Artist/Interface/Sources/Models/Artist.swift:
```diff
@@ -0,0 +1,40 @@
+import Core
+
+public struct Artist: Hashable {
+    private let uuid: UUID = .init()
```
물론 id도 서버가 문제없이 제공 해 주면 고유값이 될 수 있지만 diffableDataSource에서는 단순 id값만 객체가 들고 있으면 중복된 값이 들어올 수 있는 환경이라고 판단하여 에러를 내뱉습니다. 이런 문제를 해결하기 위해 각 Artist객체에 uuid를 부여하였습니다!

inwoodev commented on Projects/Domain/Artist/Interface/Sources/Services/ArtistPaginationService.swift:
```diff
@@ -0,0 +1,59 @@
+}
+
+final public class ArtistPaginationService: ArtistPaginationServiceInterface {
+    public var currentSearchWord: String = ""
+    public var currentPage: Int?
+    public var hasNextPage: Bool = true
+    public var isLoading: Bool = false
+    
+    private let lock = NSLock()
+    
+    public init() { }
+    
+    public func update(currentPage: Int?, hasNextPage: Bool) {
+        lock.lock()
+        defer {
+            self.lock.unlock()
+        }
+        
+        self.currentPage = currentPage
+        self.hasNextPage = hasNextPage
+    }
+    
+    public func setLoading(_ loading: Bool) {
+        lock.lock()
+        defer {
+            self.lock.unlock()
+        }
+        
+        self.isLoading = loading
+    }
+    
+    public func setCurrentSearchWord(_ word: String) {
+        lock.lock()
+        defer {
+            self.lock.unlock()
+        }
+        
+        self.currentSearchWord = word
+    }
+}
```
페이지 관리를 위해 필요한 객체입니다.

GetArtistsUseCase 그리고 SearchArtistsUsecase는 해당 객체를 공유하여 사용해서 현재 페이지 상태 및 추가 데이터를 불러올 수 있는지 등을 판단할 수 있기에 세심한 관리가 필요합니다.

각 프로퍼티들은 flag역할을 하며 비동기적인 상황에서 충분히 업데이트 될 수 있기에 NSLock을 활용하여 스레드 안전성을 보장할 수 있게 설계하였습니다~!

inwoodev commented on Projects/Feature/Main/Interface/Sources/Errors/ArtistSelectionError.swift:
```diff
@@ -0,0 +1,52 @@
+    case tooManyFavorites(limit: Int)
+    case domainError(ArtistError)
+    
+    // MARK: - 추후 기획과 논의 후 메세지 수정 필요
+    
+    public var errorDescription: String? {
+        switch self {
+        case .tooManyFavorites(let limit):
+            return "아티스트는 최대 \(limit)명까지\n 선택할 수 있어요."
+            
+        case .domainError(let artistError):
+            switch artistError {
+            case .networkError(let networkError):
+                switch networkError {
+                case .serverError, .feelinAPIError:
+                    return "서버에서 아티스트 데이터를 처리하는 과정에서 문제가 생겼어요.\n 잠시 후 다시 시도 해 주세요\n 문제가 지속적으로 발생한다면 관리자에게 문의 주세요."
+                    
+                case .unknownError(let description):
+                    return "알 수 없는 네트워크 오류가 발생했습니다: \(description)"
+                    
+                default:
+                    return "디바이스에서 아티스트 데이터를 처리하는 과정에서 문제가 생겼어요.\n 잠시 후 다시 시도 해 주세요\n 문제가 지속적으로 발생한다면 관리자에게 문의 주세요."
+                }
+            case .unknown(let errorDescription):
+                return "\(errorDescription).\n잠시 후 다시 시도 해 주세요.\n 문제가 지속적으로 발생한다면 관리자에게 문의 주세요."
+                
+            case .keychainError:
+                return "인증 처리과정에서 문제가 생겼어요. \n잠시 후 다시 시도 해 주세요.\n 문제가 지속적으로 발생한다면 관리자에게 문의 주세요."
+            }
+        }
+    }
```
유저에게 보여질 에러메세지를 일단 임시로 구현 해 보았습니다. 기획쪽과 충분히 논의 후 추후 수정해야 합니다.

inwoodev commented on Projects/Feature/Main/Example/Sources/SceneDelegate.swift:
```diff
@@ -23,7 +28,37 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         window = UIWindow(windowScene: windowScene)
         
-//        window?.rootViewController = ArtistSelectViewController()
+        DIContainer.standard.register(.networkProvider) { resolver in
+            return NetworkProvider(
+                networkSession: .init(requestInterceptor: MockTokenInterceptor())
+            )
+        }
```
앞으로 다른 화면을 개발할 때에도 대부분의 API 호출에 accessToken이 필요할 것으로 예상됩니다. 매번 실제 인증 작업을 통해 서버에서 accessToken을 가져오는 것은 비효율적이라고 생각했습니다.

따라서, 개발 중에 활용할 수 있는 accessToken을 서버에 요청하였고, 해당 토큰 정보는 저희 사이드 프로젝트인 Lyrics의 Notion APIDocs에서 확인할 수 있습니다.

이렇게 Example 앱과 MockTokenInterceptor를 통해 개발용 accessToken을 주입하면 개발 효율성을 더욱 높일 수 있을 것입니다.

inwoodev commented on Projects/Feature/Main/Testing/Sources/MockTokenInterceptor.swift:
```diff
@@ -0,0 +1,40 @@
+    public func adapt(_ urlRequest: URLRequest) -> AnyPublisher<URLRequest, Error> {
+        var request = urlRequest
+        
+        // MARK: - Example 앱에서 서버 관련 기능을 테스트할 때, DEV용 토큰을 주입하면 정상적으로 데이터를 불러올 수 있는지 확인할 수 있습니다. Bearer 뒤에 한 칸 띄우고 토큰을 입력하여 사용하세요.
+        // MARK: - [WARNING]: DEV용 토큰이지만 깃허브에는 올리지 않도록 유의해야 합니다.
+        
+        request.setValue(
+            "Bearer  <Notion → Developer → APIDocs → DEV 토큰 페이지 참고>",
+            forHTTPHeaderField: "Authorization"
+        )
```
[코멘트](https://github.com/project-lyrics/app-iOS/pull/40/commits/2e8c6f1e99336d8a088e48d953e166485043332c#r1685736791) 참고

inwoodev commented on Projects/Shared/Util/Sources/CombineCocoa/CombineControlProperty.swift:
[코멘트](https://github.com/project-lyrics/app-iOS/pull/40/commits/2f6a60a96806740bd240ed0cac99d107d2961328#r1685734231) 참고

derrickkim0109 commented on Projects/Core/Network/Interface/Sources/API/FeelinAPI.swift:
```diff
@@ -15,11 +15,14 @@ public enum FeelinAPI<R> {
     )
     case checkUserValidity
     case reissueAccessToken(refreshToken: String)
+    case getArtists(cursor: Int?, size: Int)
+    case searchArtists(query: String, cursor: Int?, size: Int)
+    case postFavoriteArtists(ids: [Int])
```
**P5**
cursor가 Optional인 이유가 있을까요?

inwoodev replied:
오 네 이게 swagger참고 해서 구현한건데요. 지금은 수정되어서 문장이 달라진것 같은데 아티스트리스트 조회 api 콜을 하는 과정에서 requrest에 cursor파라메터가 없을 경우 초기 데이터를 제공한다고 적혀있어서요. 그래서 현재는 cursor값이 nil 인 경우 request파라미터를 아래와 같이 구현합니다~

``` swift
// FeelinAPI.swift
.
.
.
    public var queryParameters: Encodable? {
        switch self {
        case .getArtists(let cursor, let size):
            if let cursor = cursor {
                return [
                    "cursor": cursor,
                    "size": size
                ]
            }
            
            return [
                "size": size
            ]
.
.
.
```

derrickkim0109 replied:
아 문서에서는 cursor 타입이 nullable하다는 표현이 없어서 여쭤봤습니다. ㅎㅎ

좋습니다~!

derrickkim0109 commented on Projects/Domain/Artist/Interface/Sources/Services/ArtistPaginationService.swift:
```diff
@@ -0,0 +1,59 @@
+
+final public class ArtistPaginationService: ArtistPaginationServiceInterface {
+    public var currentSearchWord: String = ""
+    public var currentPage: Int?
```
currentPage가 Optional인 경우가 어떤게 있을까요?

inwoodev replied:
currentPage는 cursor 용어 대신 도메인 그리고 presentation layer에 맞게 재해석한 용어입니다~

optional로 한 이유는 처음 cursor값이 1부터 시작하는지 아니면 0부터 시작하는지 확인하지 못한 부분도 있지만 (Swagger execute 이슈...) [해당 코멘트](https://github.com/project-lyrics/app-iOS/pull/40#discussion_r1688406879) 내용이 주된 이유입니다. 요건 백앤드랑 좀 얘기해봐야겠어요..분명히 cursor파라미터가 없는 케이스에 대한 내용이 swagger에 있었는데 지금 보니까 없어져 있더라구요...

derrickkim0109 replied:
이 부분은 다시 논의해보고 정확하게 정하면 좋을 것 같네요. 

조금 잘 바뀌네요..ㅎㅎ

derrickkim0109 commented on Projects/Domain/Artist/Sources/ArtistISource.swift:
```diff
@@ -0,0 +1,3 @@
+public protocol ArtistSource {
```
**P3**

파일이름과 객체이름이 다른 것 같아요~

inwoodev replied:
오 수정하겠습니다~

derrickkim0109 reviewed (commented):
너무 고생하셨습니다.

inwoodev committed:
fix: 검색 로직 수정

- 키보드가 내려가있는 상황에서 clearbutton을 탭할 경우 초기 검색결과를 불러오도록 로직 수정
- 검색 중인 상황에서 return 버튼을 탭할 경우 전체 검색이 수행되는 로직 제거

inwoodev committed:
refactor: ArtistSelectView에서 skip버튼 제거

inwoodev committed:
style: 자잘한 UI 수정

- 노래 → 곡으로 문구 수정
- FeelinArtistCell의 아티스트이름 색상 수정
- FeelinSearchBar의 돋보기 이미지 색상 수정

inwoodev commented:
@derrickkim0109  

2024.07.24 추가 수정사항
- 키보드가 내려가있는 상황에서 clearbutton을 탭할 경우 초기 검색결과를 불러오도록 로직 수정
- 검색 중인 상황에서 return 버튼을 탭할 경우 전체 검색이 수행되는 로직 제거
- ArtistSelectView에서 skip버튼 제거
- 노래 → 곡으로 문구 수정
- FeelinArtistCell의 아티스트이름 색상 수정
- FeelinSearchBar의 돋보기 이미지 색상 수정

inwoodev committed:
chore: ArtistSource 파일의 스펠링 수정
